### PR TITLE
Bnb/sup3r embedder

### PR DESCRIPTION
**Pull request by bnb32**
Added a few more obs layers, along with axial attention.

### CODE REVIEW

File: phygnn/layers/custom_layers.py [hunk 1]
```diff
@@ -943,6 +941,89 @@ def call(self, x):
...
+        ----------
+        input_shape : tuple
+            Shape tuple of the input tensor
+        """
+        self.rank = len(input_shape)
+        self.attention_layers = [
+            tf.keras.layers.MultiHeadAttention(
+                num_heads=self.num_heads, key_dim=self.key_dim
+            )
+            for _ in range(self.rank - 2)
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:984]: -2 because no attention in obs or feature axis, right? Recommend adding `assert self.rank >= 3`

File: phygnn/layers/custom_layers.py [hunk 2]
```diff
@@ -943,6 +941,89 @@ def call(self, x):
...
+        Returns
+        -------
+        x : tf.Tensor
+            Output tensor
+        """
+        x = self._apply_attention_along_axis(x, axis=1)  # Lats
+        x = self._apply_attention_along_axis(x, axis=2)  # Lons
+        if self.rank == 5:
+            x = self._apply_attention_along_axis(x, axis=3)  # Time
+        return x
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1024]: Recommend iterating `for idx, ida in enumerate(range(1, x.ndim-1))` and passing `self.attention_layers[idx]` to the apply method. Much more flexible than assuming rank==4/5 which isn't currently documented. If you do this, note somewhere that we assume ida=0 is obs and ida=-1 is features.

File: phygnn/layers/custom_layers.py [hunk 3]
```diff
@@ -1344,20 +1423,119 @@ def call(x, hi_res_feature):
...
+        """Build the weight net layer based on an input shape
+
+        Parameters
+        ----------
+        input_shape : tuple
+            Shape tuple of the input tensor
+        """
+        self.rank = len(input_shape)
+
+        if self.rank == 4:
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1472]: recommend adding assert for the ranks this can accept

File: phygnn/layers/custom_layers.py [hunk 4]
```diff
@@ -1344,20 +1423,119 @@ def call(x, hi_res_feature):
...
+            conv_class = tf.keras.layers.Conv2D
+        else:
+            conv_class = tf.keras.layers.Conv3D
+        self.embed_net = []
+        for _, filters in enumerate(self.filters):
+            self.embed_net += [
+                conv_class(
+                    filters=filters,
+                    kernel_size=1,
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1481]: kernel size 1? Isn't a conv with kernel 1 just a linear transform (scalar and adder)? If this is the case, no need for multiple?

Where are the weights for this embed network actually trained? I don't think the optimizer will see these weights.

File: phygnn/layers/custom_layers.py [hunk 5]
```diff
@@ -1344,20 +1423,119 @@ def call(x, hi_res_feature):
...
+            hi_res_feature = tf.fill(x[..., :1].shape, np.nan)
         mask = tf.math.is_nan(hi_res_feature)
         fixed = tf.where(mask, x[..., :1], hi_res_feature)
         return tf.concat((x, fixed), axis=-1)
 
 
+class Sup3rConcatEmbeddedObs(tf.keras.layers.Layer):
+    """Layer to concatenate sparse data in the middle of a super resolution
+    forward pass, with a learned embedding. This is used to condition models
+    on sparse observation data."""
```
Comment (grantbuster) [inline phygnn/layers/custom_layers.py:1444]: I'm still not convinced that this is fundamentally different from letting the the generator learn how to handle the extra layers, and i'm confused about the implementation of the embed net attribute.

**Review by grantbuster — CHANGES_REQUESTED** [review]

File: phygnn/layers/custom_layers.py (hunk 5 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1444]: This could make it easier to include the sparse data by encoding it instead of injecting physical values in latent space. I'm still seing pin points in the output and I think this will help. Also, this can encode obs within multiple channels instead of just one.

File: phygnn/layers/custom_layers.py (hunk 4 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1481]: Theres still mixing along channels, just done point wise. And non-linear mixing is done with multiple layers. This like applying multiple dense layers point wise. You scared me but the weights show up in layer.trainable_weights. Doesn't that automatically happen if the layers are defined in __init__ or build? We've defined hidden layers similarly elsewhere, no?

File: phygnn/layers/custom_layers.py (hunk 2 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1024]: Good suggestion, thanks. Think we can just pass ida to the current apply method though?

Comment (grantbuster) [reply phygnn/layers/custom_layers.py:1028]: Sure, you get my point though right? It would be nice if this was flexible enough to run on a 3D tensor with timeseries without mods.

Comment (bnb32) [reply phygnn/layers/custom_layers.py:1028]: Yeah, doesn't just passing ida the current method do that? I don't follow the suggestion to pass the layer through the method.

File: phygnn/layers/custom_layers.py (hunk 4 above)
Comment (grantbuster) [reply phygnn/layers/custom_layers.py:1488]: Okay gotcha on the activations and similar to dense layers. 

Yeah i think it needs to be in `Sup3rConcatEmbeddedObs.trainable_weights`. I'm a little surprised that the weights from an attribute show up there but if you've confirmed then should be all good. Here's an example where i think `tf.keras.layers.Layer.add_weight()` makes sure the weight is trainable, i assume something in the Conv object init does the same thing and somehow has awareness that its an attribute in a trainable layer?

https://github.com/NREL/phygnn/blob/main/phygnn/layers/custom_layers.py#L209-L215

File: phygnn/layers/custom_layers.py (hunk 5 above)
Comment (grantbuster) [reply phygnn/layers/custom_layers.py:1448]: Like i said, still not convinced :) but let me know what you find!

File: phygnn/layers/custom_layers.py (hunk 2 above)
Comment (grantbuster) [reply phygnn/layers/custom_layers.py:1028]: Yeah your new commit resolves this comment, just making sure we're on the same page on flexibility

File: phygnn/layers/custom_layers.py (hunk 4 above)
Comment (bnb32) [reply phygnn/layers/custom_layers.py:1488]: Yeah I saw that you did that with the Gaussian layer. I think when the model is initialized the layers are added to the compute graph and since conv layers already have trainable weights this works out. I think that's sort of what we assumed with these other layers, like SqueezeAndExcitation with the _hidden_layers attrs.

Comment (grantbuster) [reply phygnn/layers/custom_layers.py:1488]: Yeah gotcha. I never actually used SqueezeAndExcitation but makes sense

**Review by grantbuster — APPROVED** [review]
